dot/parachain/candidate-validation/messages.go

```diff
@@ -31,14 +32,24 @@ type ValidateFromExhaustive struct {
 }
 
 // ValidationResult represents the result coming from the candidate validation subsystem.
+// Validation results can be either Valid or Invalid.
+//
+// If the result is Invalid,
+// set the IsValid field of ValidationResultMessage to false.
+// also store the reason for invalidity in the Err field of ValidationResultMessage.
+//
+// If the result is Valid,
+// set the IsValid field of ValidationResultMessage to true.
+// set the values of the CandidateCommitments and PersistedValidationData fields of ValidationResultMessage.
+
 type ValidationResult struct {
 	IsValid                 bool
 	CandidateCommitments    parachaintypes.CandidateCommitments
 	PersistedValidationData parachaintypes.PersistedValidationData
-	Err                     error
+	ReasonForInvalidity     error
```

Comment:
Given this is the actual type:
```
pub enum ValidationResult {
	/// Candidate is valid. The validation process yields these outputs and the persisted
	/// validation data used to form inputs.
	Valid(CandidateCommitments, PersistedValidationData),
	/// Candidate is invalid.
	Invalid(InvalidCandidate),
}
```
Why are we trying to store both the `Valid(CandidateCommitments, PersistedValidationData)` tuple and `Invalid(InvalidCandidate)` enum in the same struct? Couldn't we separate it out into separate types?  I was thinking something alone the lines of 
```
type ValidationResult interface {
	isValidationResult()
}
type ValidValidationResult struct {
	CandidateCommitments    parachaintypes.CandidateCommitments
	PersistedValidationData parachaintypes.PersistedValidationData
}

func (ValidValidationResult) isValidationResult() {}

type InvalidValidationResult struct {
	InvalidCandidate
}

func (InvalidValidationResult) isValidationResult() {}

type InvalidCandidate interface {
	isInvalidCandidate()
}
type ExecutionError string

func (ExecutionError) isInvalidCandidate()

type InvalidOutputs struct{}

func (InvalidOutputs) isInvalidCandidate()

type Timeout struct{}

func (Timeout) isInvalidCandidate()
... more enum options
```
That way we're not conflating errors with reason for an invalid candidate.  You can still implement the error interface for any of the `InvalidCandidate` rust variants/interface implementers if you choose to.  Or just treat it as an `InvalidCandidate`.

It'll be easy to do a type switch between `ValidValidationResult` and `InvalidValidationResult`.

Reply:
> That way we're not conflating errors with reason for an invalid candidate

We could solve that by having an enum for Candidate Invalidity details. So, ReasonForInvalidity could be that enum instead of just error.

https://github.com/paritytech/polkadot-sdk/blob/686aa233e67c619bcdbc8b758a9ddf92c3315cf1/polkadot/node/primitives/src/lib.rs#L324


> It'll be easy to do a type switch between ValidValidationResult and InvalidValidationResult.

Agreed. We could have a type for both of those things.

I don't like the approach of having `ValidationResult` and `InvalidCandidate` interface, though. I feel that approach could bloat the code and make it less readable. We have used such approaches at places. Hopefully, we will iterate through that and replace with some better approach.

Reply:
I've updated this to add types `ValidValidationResult` and `InvalidValidatiionResult` that will be used in `ValidationResult`.  Let me know your thoughts.